File: 06_strings/06_07_valid_anagram/valid_anagram.cpp
#include <bits/stdc++.h>

using namespace std;

bool isAnagram(string s, string t)
{
    if (s.length() != t.length())
    {
        return false;
    }
    
    int count[26] = {0};

    for (char c : s)
    {
        count[c - 'a']++;
    }
    for (char c : t)
    {
        count[c - 'a']--;
    }
    for (int num : count)
    {
        if (num != 0)
        {
            return false;
        }
    }
    return true;
}

int main()
{
    // Test case 1: Basic example
    string s1 = "listen";
    string t1 = "silent";
    cout << "Test Case 1: " << (isAnagram(s1, t1) ? "true" : "false") << ", Strings: \"" << s1 << "\", \"" << t1 << "\"" << endl;

    // Test case 2: Unequal lengths
    string s2 = "hello";
    string t2 = "world";
    cout << "Test Case 2: " << (isAnagram(s2, t2) ? "true" : "false") << ", Strings: \"" << s2 << "\", \"" << t2 << "\"" << endl;

    // Test case 3: Anagram with repeated characters
    string s3 = "abcabc";
    string t3 = "bacbca";
    cout << "Test Case 3: " << (isAnagram(s3, t3) ? "true" : "false") << ", Strings: \"" << s3 << "\", \"" << t3 << "\"" << endl;

    // Test case 4: Empty strings
    string s4 = "";
    string t4 = "";
    cout << "Test Case 4: " << (isAnagram(s4, t4) ? "true" : "false") << ", Strings: \"" << s4 << "\", \"" << t4 << "\"" << endl;

    // Test case 5: Anagram with different character counts
    string s5 = "abcd";
    string t5 = "abc";
    cout << "Test Case 5: " << (isAnagram(s5, t5) ? "true" : "false") << ", Strings: \"" << s5 << "\", \"" << t5 << "\"" << endl;

    // Test case 6: Anagram with different characters
    string s6 = "abcd";
    string t6 = "efgh";
    cout << "Test Case 6: " << (isAnagram(s6, t6) ? "true" : "false") << ", Strings: \"" << s6 << "\", \"" << t6 << "\"" << endl;

    return 0;
}

File: 06_strings/06_07_valid_anagram/ValidAnagram.java
public class ValidAnagram {
    public static void main(String[] args) {
        // Test case 1: Basic example
        String s1 = "listen";
        String t1 = "silent";
        System.out.println(
                "Test Case 1: " + (isAnagram(s1, t1) ? "true" : "false") + ", Strings: \"" + s1 + "\", \"" + t1 + "\"");

        // Test case 2: Unequal lengths
        String s2 = "hello";
        String t2 = "world";
        System.out.println(
                "Test Case 2: " + (isAnagram(s2, t2) ? "true" : "false") + ", Strings: \"" + s2 + "\", \"" + t2 + "\"");

        // Test case 3: Anagram with repeated characters
        String s3 = "abcabc";
        String t3 = "bacbca";
        System.out.println(
                "Test Case 3: " + (isAnagram(s3, t3) ? "true" : "false") + ", Strings: \"" + s3 + "\", \"" + t3 + "\"");

        // Test case 4: Empty strings
        String s4 = "";
        String t4 = "";
        System.out.println(
                "Test Case 4: " + (isAnagram(s4, t4) ? "true" : "false") + ", Strings: \"" + s4 + "\", \"" + t4 + "\"");

        // Test case 5: Anagram with different character counts
        String s5 = "abcd";
        String t5 = "abc";
        System.out.println(
                "Test Case 5: " + (isAnagram(s5, t5) ? "true" : "false") + ", Strings: \"" + s5 + "\", \"" + t5 + "\"");

        // Test case 6: Anagram with different characters
        String s6 = "abcd";
        String t6 = "efgh";
        System.out.println(
                "Test Case 6: " + (isAnagram(s6, t6) ? "true" : "false") + ", Strings: \"" + s6 + "\", \"" + t6 + "\"");
    }

    private static boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }
        int[] count = new int[26];
        for (char c : s.toCharArray()) {
            count[c - 'a']++;
        }
        for (char c : t.toCharArray()) {
            count[c - 'a']--;
        }
        for (int num : count) {
            if (num != 0) {
                return false;
            }
        }
        return true;
    }
}
